Mathlib/Algebra/Lie/LieTheorem.lean

```diff
@@ -0,0 +1,554 @@
+/-
+Copyright (c) 2024 Lucas Whitfield. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Lucas Whitfield
+-/
+import Mathlib.Algebra.Lie.Solvable
+import Mathlib.Algebra.Lie.Weights.Linear
+
+/-!
+# Lie's theorem for Solvable Lie algebras.
+
+This file proves Lie's theorem, the statement that Lie modules of solvable Lie algebras over
+algebraically closed fields of characteristic 0 have a common eigenvector for the action of all
+elements of the Lie algebra. This result is named `LieModule.exists_forall_lie_eq_smul_of_Solvable`.
+-/
+
+open LieAlgebra
+
+
+-- let k be a field of characteristic zero
+variable {k : Type*} [Field k] [CharZero k] [IsAlgClosed k]
+-- Let L be a Lie algebra over k
+variable {L : Type*} [LieRing L] [LieAlgebra k L]
+-- and let V be a finite-dimensional k-representation of L
+variable {V : Type*} [AddCommGroup V] [Module k V] [FiniteDimensional k V]
+  [LieRingModule L V] [LieModule k L V] [Nontrivial V]
+
+open FiniteDimensional
+
+/- We define the Submodules generated by repeatedly applying a linear map `f: V →ₗ[F] V`
+to a vector `v`-/
+section
+
+variable {F V : Type*} [Field F] [AddCommGroup V] [Module F V] [FiniteDimensional F V]
+variable (v : V) (f : Module.End F V)
+-- the Submodule generated by `v`, `f v`, `f f v`, ... , `f^[n-1] v`
+abbrev U (n : ℕ) : Submodule F V := Submodule.span F {v' : V | ∃ a < n, v' = (f^a) v}
+
+lemma U_zero_eq_bot : U v f 0 = ⊥ := by
+  rw [Submodule.span_eq_bot]
+  intro x ⟨a, ha, _⟩
+  contradiction
+
+lemma U_mono {a b : ℕ} (h : a ≤ b) : U v f a ≤ U v f b :=
+  Submodule.span_mono (fun _ ⟨c, hc, hw⟩ ↦ ⟨c, lt_of_lt_of_le hc h, hw⟩)
+
+lemma map_U_le (n : ℕ) : Submodule.map f (U v f n) ≤ U v f (n + 1) := by
+  rw [Submodule.map_span]
+  apply Submodule.span_mono
+  intro w ⟨z, ⟨m, hm, hz⟩, hw⟩
+  use m + 1, Nat.add_lt_add_right hm 1
+  rw [← hw, hz, ← LinearMap.mul_apply, ← pow_succ']
+
+def U_map : ℕ →o Submodule F V where
+  toFun := U v f
+  monotone' _ _ := U_mono v f
+
+abbrev iSupU : Submodule F V := ⨆ k : ℕ, U_map v f k
+
+lemma iSupU_eq_U : ∃ N : ℕ, iSupU v f = U v f N := by
+  apply Submodule.FG.stabilizes_of_iSup_eq (IsNoetherian.noetherian (iSupU v f)) (U_map v f)
+  rfl
+
+lemma map_iSupU_le_iSupU: Submodule.map f (iSupU v f) ≤ iSupU v f := by
+  rcases iSupU_eq_U v f with ⟨N, hN⟩
+  nth_rewrite 1 [hN]
+  apply le_trans' $ le_iSup (U v f) (N + 1)
+  exact map_U_le v f N
+
+end
+
+section
+variable (k V)
+-- the lie action of `L` on `V`
+abbrev π (z : L) : Module.End k V := LieModule.toEnd k L V z
+variable {k V}
+
+variable (A : LieIdeal k L) (χ : Module.Dual k A)
+
+abbrev T (w : A) : Module.End k V := (π k V w)  - χ w • 1
+
+-- We define a submodule of V which is L-invariant
+def altWeightSpace : Submodule k V where
+  carrier := {v |∀ a : A, ⁅a.val,v⁆ = (χ a)•v}
```

Comment:
A comment on style: in general, we prefer to put spaces around binary operators (like the dot) or the vertical bar.
This also applies in other places, please make the same change there.
```suggestion
  carrier := {v | ∀ a : A, ⁅a.val, v⁆ = (χ a) • v}
```

(*Just in case* you haven't seen it already, mathlib's style guide is here: https://leanprover-community.github.io/contribute/style.html)